{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cache/home/as2654/Yang/crossref_annot\n",
      "\n",
      " ['esxJ_Alland.faa', 'cross_ref_genes.py', 'repair_pgap_gbk2.ipynb', 'repaired_pgap_03012022.gbff', 'AllandRv_results', '.ipynb_checkpoints', 'Rv1_09302022', 'acc_metadata_save.json', 'Rv1_09302022.zip', 'notebook.tex', 'esxN_Alland.faa', 'repair_pgap_gbk_09302022.ipynb', 'repair_pgap_gbk.ipynb', 'repaired_pgapout.gbff', 'replace_out.txt', 'GCA_000195955.2_ASM19595v2_genomic.fna', 'GCA_000195955.2_ASM19595v2_genomic.gbff', 'AllandRv_results.zip']\n",
      "\n",
      " ['cwltool.log', 'annot.gbk', 'annot.fna', 'final_asnval_diag.xml', 'initial_asnval_diag.xml', 'final_asndisc_diag.xml', 'annot.gff', 'annot.faa', 'H37Rv_Dec2019Updated_Three_pilon.fasta', 'initial_asndisc_diag.xml', 'annot.sqn', 'calls.tab']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Contact Avi D Shah at        as2654 at njms.rutgers.edu\n",
    "Contact Dr. Jason Yang at    jason.y at rutgers.edu\n",
    "\n",
    "10/1/2022\n",
    "\n",
    "The purpose of this script is to repair genbank files output by PGAP.\n",
    "\n",
    "You should first run the final cell to initialize parse_genbank.\n",
    "\n",
    "\"Repairing\" genbank files consists of replacing gene fields with \n",
    "gene names/synonyms (and locus tags when these are not available),\n",
    "and numbering repeats (e.g. esxN.k where k is the index of the gene).\n",
    "Locus tag fields will be replaced with the appropriate Rv# from the \n",
    "current Mtb reference made available by the NCBI (NC_000962.3).\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import json\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "from multiprocessing import Process, Manager\n",
    "from math import ceil\n",
    "import time\n",
    "\n",
    "print(os.getcwd())\n",
    "print(\"\\n\", os.listdir(\".\"))\n",
    "print(\"\\n\", os.listdir(\"Rv1_09302022\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# methods to parse genbank files\n",
    "\n",
    "import re\n",
    "\n",
    "def chunk_genbank(path_to_gbk):\n",
    "    parsing_mainfile = False\n",
    "    parsing_tail = False\n",
    "\n",
    "    header = []\n",
    "    contents = []\n",
    "    tail = []\n",
    "\n",
    "    with open(path_to_gbk, \"r\") as gbk_read:\n",
    "        current_record = []\n",
    "        for line in gbk_read:\n",
    "            if not parsing_mainfile and not parsing_tail:\n",
    "                header.append(line)\n",
    "                if line.startswith(\"FEATURES\"):\n",
    "                    parsing_mainfile = True\n",
    "                    continue\n",
    "            if parsing_mainfile:\n",
    "                if line.startswith(\"ORIGIN\"):\n",
    "                    contents.append(current_record)\n",
    "                    parsing_mainfile = False\n",
    "                    parsing_tail = True\n",
    "                else:\n",
    "                    # start a new record\n",
    "                    if line.startswith(\" \" * 5) and not line.startswith(\" \" * 21):\n",
    "                        if current_record:\n",
    "                            contents.append(current_record)\n",
    "                            current_record = []\n",
    "                        current_record.append(line)\n",
    "                    elif line.startswith(\" \" * 21):\n",
    "                        current_record.append(line)\n",
    "            if parsing_tail:\n",
    "                tail.append(line)\n",
    "    return header, contents, tail\n",
    "\n",
    "\n",
    "# utility methods for extracting features of interest\n",
    "# note, these require the output of parse_genbank for the\n",
    "# same file to be stored in global var \"pgap_out\"\n",
    "\n",
    "import re\n",
    "\n",
    "def get_range(content_block):\n",
    "    if type(content_block) is list:\n",
    "        return re.sub(\"\\s+\", \" \", content_block[0]).strip().split(\" \")[1]\n",
    "    elif type(content_block) is str: # this contingency can be ignored\n",
    "        return re.sub(\"\\s+\", \" \", content_block).strip().split(\" \")[1]\n",
    "    return None\n",
    "\n",
    "# added support for multiline fields like \"inference\" and \"translation\"\n",
    "def get_field(content_block, field_name, multiline_sep = \" \"):\n",
    "    field_found = False\n",
    "    retstr = \"\"\n",
    "    for line_idx in range(len(content_block)):\n",
    "        line = content_block[line_idx]\n",
    "        if field_name in line and not field_found:\n",
    "            field_found = True\n",
    "            retstr += line.strip().split(\"=\")[1]\n",
    "        elif field_found:\n",
    "            if line.strip().startswith(\"/\"):\n",
    "                return re.sub(\"\\\"\", \"\", retstr)\n",
    "            else:\n",
    "                retstr += multiline_sep + line.strip()\n",
    "    return re.sub(\"\\\"\", \"\", retstr)\n",
    "\n",
    "\n",
    "def get_inference_acc(content_block):\n",
    "    #current_range = get_range(content_block)\n",
    "    inference_val = get_field(content_block, \"inference\")\n",
    "    if \"RefSeq\" in inference_val:\n",
    "        return inference_val.split(\"RefSeq:\")[1]\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       cluster_001_consensus_pilon_pilon_pil> 4417942 bp    DNA\n",
      "            circular BCT 30-SEP-2022\n",
      "DEFINITION  Mycobacterium tuberculosis strain H37Rv chromosome, complete\n",
      "\n",
      "     source          1..4417942\n",
      "                     /organism=\"Mycobacterium tuberculosis\"\n",
      "                     /mol_type=\"genomic DNA\"\n",
      "                     /strain=\"H37Rv\"\n",
      "                     /db_xref=\"taxon:1773\"\n",
      "\n",
      "     gene            1..1524\n",
      "                     /gene=\"dnaA\"\n",
      "                     /locus_tag=\"Rv1_000001\"\n",
      "\n",
      "     CDS             1..1524\n",
      "                     /gene=\"dnaA\"\n",
      "                     /locus_tag=\"Rv1_000001\"\n",
      "                     /inference=\"COORDINATES: similar to AA\n",
      "                     sequence:RefSeq:NP_214515.1\"\n",
      "                     /note=\"Derived by automated computational analysis using\n",
      "                     gene prediction method: Protein Homology.\n",
      "                     GO_function: GO:0003688 - DNA replication origin binding\n",
      "                     [Evidence IEA];\n",
      "                     GO_function: GO:0005524 - ATP binding [Evidence IEA];\n",
      "                     GO_function: GO:0043565 - sequence-specific DNA binding\n",
      "                     [Evidence IEA];\n",
      "                     GO_process: GO:0006275 - regulation of DNA replication\n",
      "                     [Evidence IEA]\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"chromosomal replication initiator protein DnaA\"\n",
      "                     /protein_id=\"extdb:Rv1_000001\"\n",
      "                     /translation=\"MTDDPGSGFTTVWNAVVSELNGDPKVDDGPSSDANLSAPLTPQQ\n",
      "                     RAWLNLVQPLTIVEGFALLSVPSSFVQNEIERHLRAPITDALSRRLGHQIQLGVRIAP\n",
      "                     PATDEADDTTVPPSENPATTSPDTTTDNDEIDDSAAARGDNQHSWPSYFTERPHNTDS\n",
      "                     ATAGVTSLNRRYTFDTFVIGASNRFAHAAALAIAEAPARAYNPLFIWGESGLGKTHLL\n",
      "                     HAAGNYAQRLFPGMRVKYVSTEEFTNDFINSLRDDRKVAFKRSYRDVDVLLVDDIQFI\n",
      "                     EGKEGIQEEFFHTFNTLHNANKQIVISSDRPPKQLATLEDRLRTRFEWGLITDVQPPE\n",
      "                     LETRIAILRKKAQMERLAVPDDVLELIASSIERNIRELEGALIRVTAFASLNKTPIDK\n",
      "                     ALAEIVLRDLIADANTMQISAATIMAATAEYFDTTVEELRGPGKTRALAQSRQIAMYL\n",
      "                     CRELTDLSLPKIGQAFGRDHTTVMYAQRKILSEMAERREVFDHVKELTTRIRQRSKR\"\n",
      "ORIGIN      \n",
      "        1 ttgaccgatg accccggttc aggcttcacc acagtgtgga acgcggtcgt ctccgaactt\n",
      "       61 aacggcgacc ctaaggttga cgacggaccc agcagtgatg ctaatctcag cgctccgctg\n"
     ]
    }
   ],
   "source": [
    "# get pgap results\n",
    "outpath_gbk = os.path.join(\"Rv1_09302022\", \"annot.gbk\")\n",
    "\n",
    "header   = []\n",
    "contents = []\n",
    "tail     = []\n",
    "\n",
    "header, contents, tail = chunk_genbank(outpath_gbk)\n",
    "\n",
    "# print out first few entries, see that parser worked\n",
    "for hl in header[:3]:\n",
    "    print(hl[:-1])\n",
    "for content_group in contents[:3]:\n",
    "    print()\n",
    "    for cgl in content_group:\n",
    "        print(cgl[:-1])\n",
    "for tl in tail[:3]:\n",
    "    print(tl[:-1])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rv1_000001\n",
      "NP_214515.1\n",
      "Rv1_000002\n",
      "NP_214516.1\n",
      "Rv1_000003\n",
      "NP_214517.1\n",
      "Rv1_000004\n",
      "NP_214518.1\n",
      "Rv1_000005\n",
      "NP_214519.2\n",
      "Rv1_000006\n",
      "NP_214520.1\n",
      "Rv1_000007\n",
      "NP_214521.1\n",
      "Rv1_000008\n",
      "None\n",
      "Rv1_000009\n",
      "None\n",
      "Rv1_000010\n",
      "WP_003400294.1\n",
      "Rv1_000011\n",
      "None\n",
      "Rv1_000012\n",
      "NP_214522.1\n",
      "Rv1_000013\n",
      "WP_003400321.1\n",
      "Rv1_000014\n",
      "NP_214524.1\n",
      "Rv1_000015\n",
      "NP_214525.1\n",
      "Rv1_000016\n",
      "WP_003872099.1\n",
      "Rv1_000017\n",
      "YP_177615.1\n",
      "Rv1_000018\n",
      "NP_214528.1\n",
      "Rv1_000019\n",
      "NP_214529.1\n",
      "Rv1_000020\n",
      "NP_214530.1\n",
      "Rv1_000021\n",
      "NP_214531.1\n",
      "Rv1_000022\n",
      "WP_015385224.1\n",
      "Rv1_000023\n",
      "NP_214533.1\n",
      "Rv1_000024\n",
      "NP_214534.1\n",
      "Rv1_000025\n",
      "None\n",
      "Rv1_000026\n",
      "NP_214535.1\n",
      "Rv1_000027\n",
      "NP_214536.1\n",
      "Rv1_000028\n",
      "NP_214537.1\n",
      "Rv1_000029\n",
      "NP_214538.1\n",
      "Rv1_000030\n",
      "NP_214539.1\n"
     ]
    }
   ],
   "source": [
    "# associate Rv1_00xxxx locus tags with refseq inference accessions\n",
    "\n",
    "ltags_Rv1 = []\n",
    "\n",
    "# list all Rv1_00XXXX locus tags in ascending order\n",
    "# (this *should* just be Rv_00001 to Rv_00xxxx (the highest)\n",
    "# without any gaps)\n",
    "for content_group in contents:\n",
    "    ltag = get_field(content_group, \"/locus_tag\")\n",
    "    if ltag and ltag not in ltags_Rv1:\n",
    "        ltags_Rv1.append(ltag)\n",
    "\n",
    "\n",
    "# create an association between locus_tags and RefSeq inference accessions\n",
    "ltag_acc_map = {}\n",
    "for content_group in contents:\n",
    "    ltag = get_field(content_group, \"/locus_tag\")\n",
    "    acc = get_inference_acc(content_group)\n",
    "    ltag_acc_map[ltag] = acc\n",
    "\n",
    "# see the first 30 locus tags and their RefSeq inference accs\n",
    "for l in ltags_Rv1[:30]:\n",
    "    print(l)\n",
    "    print(ltag_acc_map[l])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "# get info from NCBI; gene names and legacy locus tags for each\n",
    "# inference RefSeq accession\n",
    "\n",
    "Entrez.email = \"as2654@njms.rutgers.edu\"\n",
    "\n",
    "handle = Entrez.efetch(db=\"protein\", id=\"NP_214536.1\", retmode=\"xml\")\n",
    "record = Entrez.read(handle)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "\n",
      "\n",
      "\n",
      "GBQualifier_name organism\n",
      "GBQualifier_value Mycobacterium tuberculosis H37Rv\n",
      "\n",
      "GBQualifier_name strain\n",
      "GBQualifier_value H37Rv\n",
      "\n",
      "GBQualifier_name type_material\n",
      "GBQualifier_value type strain of Mycobacterium tuberculosis\n",
      "\n",
      "GBQualifier_name db_xref\n",
      "GBQualifier_value taxon:83332\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GBQualifier_name product\n",
      "GBQualifier_value transcriptional regulator WhiB5\n",
      "\n",
      "GBQualifier_name calculated_mol_wt\n",
      "GBQualifier_value 15047\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GBQualifier_name region_name\n",
      "GBQualifier_value Whib\n",
      "\n",
      "GBQualifier_name note\n",
      "GBQualifier_value Transcription factor WhiB; pfam02467\n",
      "\n",
      "GBQualifier_name db_xref\n",
      "GBQualifier_value CDD:426788\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GBQualifier_name gene\n",
      "GBQualifier_value whiB5\n",
      "\n",
      "GBQualifier_name locus_tag\n",
      "GBQualifier_value Rv0022c\n",
      "\n",
      "GBQualifier_name gene_synonym\n",
      "GBQualifier_value whmG\n",
      "\n",
      "GBQualifier_name coded_by\n",
      "GBQualifier_value complement(NC_000962.3:27023..27442)\n",
      "\n",
      "GBQualifier_name experiment\n",
      "GBQualifier_value DESCRIPTION:Gene expression, mutation analysis[PMID:22733573]\n",
      "\n",
      "GBQualifier_name experiment\n",
      "GBQualifier_value EXISTENCE:Mass spectrometry[PMID:15525680]\n",
      "\n",
      "GBQualifier_name note\n",
      "GBQualifier_value alternate gene name: whmG, WhiB-like regulatory protein.\n",
      "\n",
      "GBQualifier_name transl_table\n",
      "GBQualifier_value 11\n",
      "\n",
      "GBQualifier_name db_xref\n",
      "GBQualifier_value GeneID:887071\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "save = False\n",
    "print(len(record[0][\"GBSeq_feature-table\"]))\n",
    "for val in record[0][\"GBSeq_feature-table\"]:\n",
    "    for k, v in val.items():\n",
    "        if k == \"GBFeature_quals\":\n",
    "            for a in v:\n",
    "                for b, c in a.items():\n",
    "                    print(b, c)\n",
    "                print(\"\")\n",
    "        print(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
